eager quant: remove fake_quant after add/mul nodes during QAT

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49420 fx quant: make sure observer is inserted before a quantized output
* #49382 fx quant: fix fq when input is quantized and node does not need fq
* #49239 fx quant: do not insert observers at quantized inputs
* #49238 fx quant: move {input|output}_quantized_idxs cfg from convert to prepare
* **#49213 eager quant: remove fake_quant after add/mul nodes during QAT**

Summary:

Changes behavior of Eager mode quantization to remove observation after `add_scalar/mul_scalar`.
This is not used, and it removes one difference between Eager and FX modes.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quantized_add_qat
python test/test_quantization.py TestQuantizeFxOps.test_quantized_mul_qat
python test/test_quantization.py TestQuantizationAwareTraining.test_add_scalar_uses_input_qparams
python test/test_quantization.py TestQuantizationAwareTraining.test_mul_scalar_uses_input_qparams
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25486276](https://our.internmc.facebook.com/intern/diff/D25486276)

vkuzo committed:
fx quant: insert fake_quant after add/mul nodes during QAT

Summary:

Changes behavior of FX graph mode quantization to insert a fake_quant
after add and mul nodes.  This is useful to model numerics better,
since these nodes will be quantized during inference, and
adding/multiplying by a scalar can result in values in between
quantization bins during training.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quantized_add_qat
python test/test_quantization.py TestQuantizeFxOps.test_quantized_mul_qat
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -197,16 +197,27 @@ def is_observed(input_arg):
+            else:
+                # for QAT, we always insert a fake_quant after add/mul
+                new_observer = qconfig.activation()
+                insert_observer(
```
I think we'll need to use the same fake quantize module as the input here to properly simulate quantization since output of quantized::add with scalar input inherits quantization parameters from input

raghuramank100 replied:
Not sure I follow: The current logic inserts observers as long as one of the inputs is a tensor. why is this a problem for QAT?

jerryzh168 commented:
we can match eager now and improve later. since eager mode is incorrect, because the fake quantize is ignored, and the model before and after convert for the qat model would have different numerics. output of add_scalar/mul_scalar need to use the same fake quantize module as input so that the numerics of fake quantize matches the quantized op.

vkuzo committed:
Update on "fx quant: insert fake_quant after add/mul nodes during QAT"


Summary:

Changes behavior of FX graph mode quantization to insert a fake_quant
after add and mul nodes.  This is useful to model numerics better,
since these nodes will be quantized during inference, and
adding/multiplying by a scalar can result in values in between
quantization bins during training.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quantized_add_qat
python test/test_quantization.py TestQuantizeFxOps.test_quantized_mul_qat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25486276](https://our.internmc.facebook.com/intern/diff/D25486276)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1458,6 +1458,57 @@ def test_quantized_mul_relu(self):
+                x = self.conv1(x)
+                x = torch.add(x, 1.0)
+                x = torch.relu(x)
+                x = self.conv2(x)
```
Is this 4 because of add relu fusion?

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1458,6 +1458,57 @@ def test_quantized_mul_relu(self):
+                self.conv2 = torch.nn.Conv2d(1, 1, 1)
+
+            def forward(self, x):
+                x = torch.add(x, 1.0)
```
Are the inputs and outputs quantized? In this case, shouldnt we have 5 fake-quants? (add, conv1, add-relu, conv2, output)?

jerryzh168 replied:
output of add/input of conv1 is considered to be observed here I think

jerryzh168 replied:
the ideal state would be the output of add/input of conv1 being fake quantized with the same fq module as the input of add.

jerryzh168 commented:
looks like this breaks some of the numeric_suite tests. we may need to keep the `activation_post_process` attribute (and just use Identity).

vkuzo committed:
comments on "eager quant: remove fake_quant after add/mul nodes during QAT"


Summary:

Changes behavior of Eager mode quantization to remove observation after `add_scalar/mul_scalar`.
This is not used, and it removes one difference between Eager and FX modes.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quantized_add_qat
python test/test_quantization.py TestQuantizeFxOps.test_quantized_mul_qat
python test/test_quantization.py TestQuantizationAwareTraining.test_add_scalar_uses_input_qparams
python test/test_quantization.py TestQuantizationAwareTraining.test_mul_scalar_uses_input_qparams
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25486276](https://our.internmc.facebook.com/intern/diff/D25486276)

[ghstack-poisoned]

vkuzo commented:
> looks like this breaks some of the numeric_suite tests. we may need to keep the activation_post_process attribute (and just use Identity).

thanks, the fix was to do the same change on `QFunctional`.  `add_scalar` and `mul_scalar` will no longer be supported by Eager NS after this PR, which is intended as Eager NS only works on activations.  Theoretically someone could hack it back in with a module swap, but that should probably be outside of this PR.

vkuzo committed:
rebase on "eager quant: remove fake_quant after add/mul nodes during QAT"


Summary:

Changes behavior of Eager mode quantization to remove observation after `add_scalar/mul_scalar`.
This is not used, and it removes one difference between Eager and FX modes.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quantized_add_qat
python test/test_quantization.py TestQuantizeFxOps.test_quantized_mul_qat
python test/test_quantization.py TestQuantizationAwareTraining.test_add_scalar_uses_input_qparams
python test/test_quantization.py TestQuantizationAwareTraining.test_mul_scalar_uses_input_qparams
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25486276](https://our.internmc.facebook.com/intern/diff/D25486276)

[ghstack-poisoned]
